sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/client/DefaultHttpClient.java

```diff
@@ -202,30 +205,53 @@ private void sendBody(HttpRequest httpRequest, HttpURLConnection connection) {
      * @return A HttpResponse object
      */
     private Response<?> receiveResponse(HttpRequest httpRequest, HttpURLConnection connection) {
-        try {
-            int responseCode = connection.getResponseCode();
-            HttpHeaders responseHeaders = getResponseHeaders(connection);
+        HttpHeaders responseHeaders = getResponseHeaders(connection);
+        HttpResponse<?> httpResponse = createHttpResponse(httpRequest, connection);
 
+        if (connection.getErrorStream() == null && isTextEventStream(responseHeaders)) {
             ServerSentEventListener listener = httpRequest.getServerSentEventListener();
-            if (connection.getErrorStream() == null && isTextEventStream(responseHeaders)) {
-                if (listener != null) {
-                    processTextEventStream(httpRequest, connection, listener);
-                } else {
-                    LOGGER.atInfo().log("No listener attached to the server sent "
-                        + "event http request. Treating response as regular response.");
-                }
 
-                return new HttpResponse<>(httpRequest, responseCode, responseHeaders, null);
+            if (listener != null) {
+                processTextEventStream(httpRequest, connection, listener);
             } else {
-                AccessibleByteArrayOutputStream outputStream = getAccessibleByteArrayOutputStream(connection);
-
-                return new HttpResponse<>(httpRequest, responseCode, responseHeaders,
-                    BinaryData.fromByteBuffer(outputStream.toByteBuffer()));
+                LOGGER.atInfo().log("No listener attached to the server sent event http request. Treating response as "
+                    + "regular response.");
             }
+        } else {
+            HttpRequestMetadata metadata = httpRequest.getMetadata();
+
+            if (metadata != null && metadata.getResponseBodyHandling() != null) {
+                switch (metadata.getResponseBodyHandling()) {
+                    case IGNORE:
+                        ignoreResponseBody(httpResponse, connection);
+
+                        break;
+                    case STREAM:
+                        setStreamingResponseBody(httpResponse, connection);
+
+                        break;
+                    case BUFFER:
+                    case DESERIALIZE:
+                    default:
+                        eagerlyBufferResponseBody(httpResponse, connection);
+                }
+            } else {
+                // The metadata can only be null if intentionally set through HttpRequest.setMetadata()
+                // TODO (vcolin7): Should we ensure the metadata is not null when users call HttpRequest.setMetadata()?
```

Comment:
I think it would be better to safe-check in the setter that this cannot be null.